nexus/reconfigurator/blippy/src/checks.rs

```diff
@@ -0,0 +1,524 @@
+                            },
+                        );
+                    }
+                }
+            }
+            None => {
+                // TODO-john Add a Severity::BackwardsCompatibility and note the
+                // missing filesystem pool
+            }
+        }
+
+        if let Some(dataset) = zone_config.zone_type.durable_dataset() {
+            match sled_datasets
+                .get(&dataset.dataset.pool_name.id())
+                .and_then(|by_zpool| by_zpool.get(&dataset.kind))
+            {
+                Some(dataset) => {
+                    expected_datasets.insert(dataset.id);
+                }
+                None => {
+                    blippy.push_note(
+                        Component::Sled(sled_id),
+                        Severity::Fatal,
+                        Kind::ZoneMissingDurableDataset { zone: zone_config },
+                    );
+                }
+            }
+        }
+    }
+
+    // TODO-correctness We currently only include in-service disks in the
+    // blueprint; once we include expunged or decommissioned disks too, we
+    // should filter here to only in-service.
+    let in_service_sled_zpools = blippy
+        .blueprint()
+        .blueprint_disks
+        .iter()
+        .map(|(sled_id, disk_config)| {
+            (
+                sled_id,
+                disk_config
+                    .disks
+                    .iter()
+                    .map(|disk| disk.pool_id)
+                    .collect::<BTreeSet<_>>(),
+            )
+        })
+        .collect::<BTreeMap<_, _>>();
+    let mut noted_sleds_without_disks = BTreeSet::new();
+
+    // All datasets should be on zpools that have disk records, and all datasets
+    // should have been referenced by either a zone or a disk above.
+    for (sled_id, dataset) in blippy
+        .blueprint()
+        .all_omicron_datasets(BlueprintDatasetFilter::InService)
+    {
+        if !expected_datasets.contains(&dataset.id) {
+            blippy.push_note(
+                Component::Sled(sled_id),
+                Severity::Fatal,
+                Kind::OrphanedDataset { dataset },
+            );
+            continue;
+        }
+
+        let Some(sled_zpools) = in_service_sled_zpools.get(&sled_id) else {
+            if noted_sleds_without_disks.insert(sled_id) {
+                blippy.push_note(
+                    Component::Sled(sled_id),
+                    Severity::Fatal,
+                    Kind::SledMissingDisks {
+                        why: "sled has in-service datasets",
+                    },
+                );
+            }
+            continue;
+        };
+
+        if !sled_zpools.contains(&dataset.pool.id()) {
+            blippy.push_note(
+                Component::Sled(sled_id),
+                Severity::Fatal,
+                Kind::DatasetOnNonexistentZpool { dataset },
+            );
+            continue;
+        }
+    }
+}
```

Comment:
I don't *think* this needs to hold up this PR, but WDYT about using the example system blueprint and manually altering it to validate that these conditions are actually firing?

(Seems important to test that blippy throws errors when we expect it should)

Reply:
Yes, that definitely needs to happen. I probably will do it on this PR, since it's pretty low-risk and doesn't need much in terms of review.